File: src/dart/test/fence_adversarial_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:ores_locks_and_leases/ores_locks_and_leases.dart';
import 'package:test/test.dart';

Map<String, dynamic> _map(Object? value) =>
    (value as Map<Object?, Object?>).cast<String, dynamic>();

FencedWriteRequest _request(Map<String, dynamic> value) => FencedWriteRequest(
      tenantScope: value['tenantScope'] as String,
      resourceKey: LockKey(value['resourceKey'] as String),
      fencingToken: FencingTokenText.parse(value['fencingToken'] as String),
      operationId: value['operationId'] as String,
      payloadSha256: value['payloadSha256'] as String,
      holder: value.containsKey('holder') ? value['holder'] as String : null,
      leaseId: value.containsKey('leaseId') ? value['leaseId'] as String : null,
    );

FenceWatermark _watermark(Map<String, dynamic> value) =>
    FenceWatermark.fromRequest(_request(value));

void _expectFenceError(
  void Function() callback,
  String code,
  String name,
) {
  expect(
    callback,
    throwsA(
      isA<FenceValidationException>().having(
        (error) => error.code,
        'code',
        code,
      ),
    ),
    reason: name,
  );
}

void main() {
  final corpus = _map(jsonDecode(
    File(
      Platform.environment['ORES_FENCE_ADVERSARIAL_CORPUS'] ??
          '../../conformance/cases/fence-adversarial.json',
    ).readAsStringSync(),
  ));

  test('adversarial corpus has the pinned reproducible identity', () {
    expect(corpus['schema'], 'ores.locks.fence-adversarial/v1');
    expect(corpus['generator'], 'splitmix64-v1');
    expect(corpus['seed'], '0x4f5245534c4f434b');
  });

  for (final rawCase in corpus['tokenCases'] as List<dynamic>) {
    final fixture = _map(rawCase);
    final expected = _map(fixture['expected']);
    test('adversarial token: ${fixture['name']}', () {
      final value = fixture['value'] as String;
      if (expected['ok'] as bool) {
        expect(
          FencingTokenText.parse(value).value,
          expected['canonical'],
        );
        return;
      }
      _expectFenceError(
        () => FencingTokenText.parse(value),
        expected['error'] as String,
        fixture['name'] as String,
      );
    });
  }

  for (final rawCase in corpus['requestCases'] as List<dynamic>) {
    final fixture = _map(rawCase);
    final expected = _map(fixture['expected']);
    test('adversarial request: ${fixture['name']}', () {
      if (expected['ok'] as bool) {
        final request = _request(_map(fixture['incoming']));
        expect(request.fencingToken.value, fixture['incoming']['fencingToken']);
        return;
      }
      _expectFenceError(
        () => _request(_map(fixture['incoming'])),
        expected['error'] as String,
        fixture['name'] as String,
      );
    });
  }

  for (final rawCase in corpus['decisionCases'] as List<dynamic>) {
    final fixture = _map(rawCase);
    test('adversarial decision: ${fixture['name']}', () {
      final incoming = _request(_map(fixture['incoming']));
      final current = fixture['current'] == null
          ? null
          : _watermark(_map(fixture['current']));
      final expectedError = fixture['expectedError'] as String?;
      if (expectedError != null) {
        _expectFenceError(
          () => evaluateFence(current, incoming),
          expectedError,
          fixture['name'] as String,
        );
        return;
      }

      final expected = _map(fixture['expected']);
      final decision = evaluateFence(current, incoming);
      expect(decision.kind.wire, expected['kind']);
      expect(decision.shouldApply, expected['shouldApply']);
      expect(decision.incomingToken.value, expected['incomingToken']);
      expect(decision.currentToken.value, expected['currentToken']);
      expect(decision.previousToken?.value, expected['previousToken']);
    });
  }
}

File: src/ts/test/fence-adversarial.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { test } from "node:test";

import {
  FenceValidationError,
  evaluateFence,
  fenceWatermark,
  fencedWriteRequest,
  fencingTokenText,
} from "../dist/index.js";

const corpus = JSON.parse(
  await readFile(
    process.env["ORES_FENCE_ADVERSARIAL_CORPUS"] ??
      new URL("../../../conformance/cases/fence-adversarial.json", import.meta.url),
    "utf8",
  ),
);

function assertFenceError(callback, expected, name) {
  assert.throws(
    callback,
    (error) =>
      error instanceof FenceValidationError && error.code === expected,
    name,
  );
}

test("adversarial corpus has the pinned reproducible identity", () => {
  assert.equal(corpus.schema, "ores.locks.fence-adversarial/v1");
  assert.equal(corpus.generator, "splitmix64-v1");
  assert.equal(corpus.seed, "0x4f5245534c4f434b");
});

for (const fixture of corpus.tokenCases) {
  test(`adversarial token: ${fixture.name}`, () => {
    if (fixture.expected.ok) {
      assert.equal(fencingTokenText(fixture.value), fixture.expected.canonical);
      return;
    }
    assertFenceError(
      () => fencingTokenText(fixture.value),
      fixture.expected.error,
      fixture.name,
    );
  });
}

for (const fixture of corpus.requestCases) {
  test(`adversarial request: ${fixture.name}`, () => {
    if (fixture.expected.ok) {
      const request = fencedWriteRequest(fixture.incoming);
      assert.equal(request.fencingToken, fixture.incoming.fencingToken);
      return;
    }
    assertFenceError(
      () => fencedWriteRequest(fixture.incoming),
      fixture.expected.error,
      fixture.name,
    );
  });
}

for (const fixture of corpus.decisionCases) {
  test(`adversarial decision: ${fixture.name}`, () => {
    const incoming = fencedWriteRequest(fixture.incoming);
    const current =
      fixture.current === null ? null : fenceWatermark(fixture.current);
    if (fixture.expectedError !== undefined) {
      assertFenceError(
        () => evaluateFence(current, incoming),
        fixture.expectedError,
        fixture.name,
      );
      return;
    }
    const decision = evaluateFence(current, incoming);
    assert.deepEqual(
      {
        kind: decision.kind,
        shouldApply: decision.shouldApply,
        incomingToken: decision.incomingToken,
        currentToken: decision.currentToken,
        previousToken: decision.previousToken ?? null,
      },
      fixture.expected,
      fixture.name,
    );
  });
}

test("generated wire adversaries cannot trigger coercion or escape the closed object", () => {
  let calls = 0;
  const hostile = {
    toString() {
      calls += 1;
      throw new Error("coercion executed");
    },
    [Symbol.toPrimitive]() {
      calls += 1;
      throw new Error("primitive conversion executed");
    },
  };
  const valid = corpus.requestCases.find((item) => item.expected.ok).incoming;
  for (const [field, value, code] of [
    ["tenantScope", hostile, "invalid_type"],
    ["fencingToken", hostile, "invalid_fencing_token"],
    ["payloadSha256", hostile, "invalid_payload_sha256"],
  ]) {
    assertFenceError(
      () => fencedWriteRequest({ ...valid, [field]: value }),
      code,
      field,
    );
  }
  assertFenceError(
    () => fencedWriteRequest({ ...valid, outboxKey: "forged" }),
    "unexpected_field",
    "closed object",
  );
  assert.equal(calls, 0);
});
